src/core/a-entity.js

```diff
@@ -23,155 +22,161 @@ var ONCE = {once: true};
-      this.isEntity = true;
-      this.isPlaying = false;
-      this.object3D = new THREE.Group();
-      this.object3D.el = this;
-      this.object3DMap = {};
-      this.parentEl = null;
-      this.rotationObj = {};
-      this.states = [];
-    }
-  },
+class AEntity extends ANode {
+  constructor () {
+    super();
+    this.components = {};
+    // To avoid double initializations and infinite loops.
+    this.initializingComponents = {};
+    this.componentsToUpdate = {};
+    this.isEntity = true;
+    this.isPlaying = false;
+    this.object3D = new THREE.Group();
+    this.object3D.el = this;
+    this.object3DMap = {};
+    this.parentEl = null;
+    this.rotationObj = {};
+    this.states = [];
+  }
 
   /**
    * Handle changes coming from the browser DOM inspector.
    */
-  attributeChangedCallback: {
-    value: function (attr, oldVal, newVal) {
-      var component = this.components[attr];
-      // If the empty string is passed by the component initialization
-      // logic we ignore the component update.
-      if (component && component.justInitialized && newVal === '') {
-        delete component.justInitialized;
-        return;
-      }
-      // When a component is removed after calling el.removeAttribute('material')
-      if (!component && newVal === null) { return; }
-      this.setEntityAttribute(attr, oldVal, newVal);
+  attributeChangedCallback (attr, oldVal, newVal) {
+    var component = this.components[attr];
+
+    super.attributeChangedCallback();
+    // If the empty string is passed by the component initialization
+    // logic we ignore the component update.
+    if (component && component.justInitialized && newVal === '') {
+      delete component.justInitialized;
+      return;
     }
-  },
+    // When a component is removed after calling el.removeAttribute('material')
+    if (!component && newVal === null) { return; }
+    this.setEntityAttribute(attr, oldVal, newVal);
+  }
 
   /**
-   * Add to parent, load, play.
-   */
-  attachedCallback: {
-    value: function () {
-      var assetsEl;  // Asset management system element.
-      var sceneEl = this.sceneEl;
-      var self = this;  // Component.
+  * Add to parent, load, play.
+  */
+  connectedCallback () {
+    // Defer if DOM is not ready.
+    if (document.readyState === 'loading') {
+      document.addEventListener('DOMContentLoaded', this.connectedCallback.bind(this));
+      return;
+    }
 
-      this.addToParent();
+    AEntity.prototype.doConnectedCallback.call(this);
+  }
 
-      // Don't .load() scene on attachedCallback.
-      if (this.isScene) { return; }
+  doConnectedCallback () {
+    var assetsEl;  // Asset management system element.
+    var sceneEl = this.sceneEl;
+    var self = this;  // Component.
 
-      // Gracefully not error when outside of <a-scene> (e.g., tests).
-      if (!sceneEl) {
-        this.load();
-        return;
-      }
+    // ANode method.
+    super.connectedCallback();
```

Comment:
This line should really be at the top.

Reply:
@dmarcos there is actually no issue here, but a rather confusing code. I actually didn't see you do `this.sceneEl` twice here. You do this here:
```
var sceneEl = this.sceneEl; // but we know it's actually undefined because we didn't call super.connectedCallback()
...
super.connectedCallback();
...
sceneEl = this.sceneEl;
```
so I'd suggest just to use
```
var sceneEl;
...
super.connectedCallback();
...
sceneEl = this.sceneEl;
```
or 
```
super.connectedCallback();
var sceneEl = this.sceneEl;
```

Reply:
I see you chose the first suggestion, great.